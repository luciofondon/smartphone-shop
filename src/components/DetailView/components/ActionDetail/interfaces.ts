import { SelectChangeEvent } from '@mui/material';
import { IProductAdd } from '../../../Root/interfaces';
import { IProductDetail } from '../../interfaces';

export interface IActionDetail {
  product: IProductDetail;
  handleClickAddProduct: (product: IProductAdd) => void;
}

export interface IUseActionDetail {
  product: IProductDetail;
  handleClickAddProduct: (product: IProductAdd) => void;
}

export interface IUseActionDetailResponse {
  states: {
    storage: number;
    color: number;
  };
  actions: {
    handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
    handleChange: (event: SelectChangeEvent) => void;
  };
}
