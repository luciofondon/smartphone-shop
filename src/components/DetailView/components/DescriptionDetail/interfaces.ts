import { IProductDetail } from '../../interfaces';

export interface IDescriptionDetail {
  product: IProductDetail;
}
