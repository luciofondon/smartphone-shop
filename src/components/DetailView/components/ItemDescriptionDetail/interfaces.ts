export interface IItemDescriptrionDetail {
  label: string;
  value: string;
}
