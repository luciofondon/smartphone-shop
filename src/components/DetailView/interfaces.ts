import { IProductAdd } from '../Root/interfaces';

interface IColor {
  code: number;
  name: string;
}

interface IStorage {
  code: number;
  name: string;
}

export interface IProductDetail {
  announced: string;
  audioJack: string;
  battery: string;
  bluetooth: string[];
  brand: string;
  chipset: string;
  colors: string[];
  cpu: string;
  dimentions: string;
  displayResolution: string;
  displaySize: string;
  displayType: string;
  edge: string;
  externalMemory: string;
  gprs: string;
  gps: string;
  gpu: string;
  id: string;
  imgUrl: string;
  internalMemory: string[];
  model: string;
  networkSpeed: string;
  networkTechnology: string;
  nfc: string;
  options: { colors: IColor[]; storages: IStorage[] };
  os: string;
  price: string;
  primaryCamera: string[];
  radio: string;
  ram: string;
  secondaryCmera: string[];
  sensors: string[];
  sim: string;
  speaker: string;
  status: string;
  usb: string;
  weight: string;
  wlan: string[];
}

export interface IUseDetailViewResponse {
  states: {
    product: IProductDetail | null;
  };
  actions: {
    handleClickAddProduct: (product: IProductAdd) => void;
  };
}
