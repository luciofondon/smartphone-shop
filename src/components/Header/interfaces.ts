export interface IHeader {
  totalProducts: number;
}
