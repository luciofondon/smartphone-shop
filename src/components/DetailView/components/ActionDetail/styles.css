.form-select {
    float: left;
    margin-right: 10px;
    width: 150px;
}

.action-form {
    padding-top: 20px;
}