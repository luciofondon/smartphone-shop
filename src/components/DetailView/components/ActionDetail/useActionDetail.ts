import { SelectChangeEvent } from '@mui/material';
import { useEffect, useState } from 'react';

import { IUseActionDetail, IUseActionDetailResponse } from './interfaces';

const EMPTY = -1;

const UseActionDetail = ({
  product,
  handleClickAddProduct,
}: IUseActionDetail): IUseActionDetailResponse => {
  const [storage, setStorage] = useState<number>(EMPTY);
  const [color, setColor] = useState<number>(EMPTY);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    handleClickAddProduct({ id: product.id, colorCode: color, storageCode: storage });
  };

  const handleChange = (event: SelectChangeEvent) => {
    if (event.target.name == 'color') {
      setColor(Number(event.target.value));
    } else if (event.target.name == 'storage') {
      setStorage(Number(event.target.value));
    }
  };

  useEffect(() => {
    if (product) {
      setColor(product.options.colors[0].code);
      setStorage(product.options.storages[0].code);
    }
  }, [product]);

  return {
    states: {
      storage,
      color,
    },
    actions: {
      handleSubmit,
      handleChange,
    },
  };
};

export default UseActionDetail;
