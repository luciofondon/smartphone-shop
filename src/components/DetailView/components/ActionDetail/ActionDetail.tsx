import React from 'react';
import { Select, MenuItem, Button, InputLabel, FormControl } from '@mui/material';

import { IActionDetail } from './interfaces';
import './styles.css';
import useActionDetail from './useActionDetail';

const ActionDetail = ({ product, handleClickAddProduct }: IActionDetail) => {
  const {
    states: { storage, color },
    actions: { handleSubmit, handleChange },
  } = useActionDetail({ product, handleClickAddProduct });

  return (
    <div className="action-form">
      <form onSubmit={handleSubmit}>
        <div className="form-select">
          <FormControl fullWidth>
            <InputLabel id="storage-label">Almacenamiento</InputLabel>
            <Select
              labelId="storage-label"
              label="Almacenamiento"
              name="storage"
              value={storage.toString()}
              onChange={handleChange}
            >
              {product.options.storages.map((storage) => (
                <MenuItem value={storage.code} key={storage.code}>
                  {storage.name}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </div>

        <div className="form-select">
          <FormControl fullWidth>
            <InputLabel id="color-label">Color</InputLabel>
            <Select
              labelId="color-label"
              label="Color"
              name="color"
              value={color.toString()}
              onChange={handleChange}
            >
              {product.options.colors.map((color) => (
                <MenuItem value={color.code} key={color.code}>
                  {color.name}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </div>
        <Button type="submit">AÃ±adir a la cesta</Button>
      </form>
    </div>
  );
};

export default ActionDetail;
