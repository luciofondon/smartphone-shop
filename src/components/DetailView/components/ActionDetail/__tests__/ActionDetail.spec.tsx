import React from 'react';
import { render } from '@testing-library/react';
import ActionDetail from '../ActionDetail';
import { IProductDetail } from '../../../interfaces';

test('render ActionDetail', () => {
  const productMock: IProductDetail = {
    announced: '',
    audioJack: '',
    battery: 'Removable Li-Po 4080 mAh battery',
    bluetooth: [],
    brand: 'Alcatel',
    chipset: '',
    colors: [],
    cpu: 'Octa-core 1.3 GHz Cortex-A53',
    dimentions: '153.8 x 75.6 x 8.5 mm (6.06 x 2.98 x 0.33 in)',
    displayResolution: '5.5 inches ',
    displaySize: '',
    displayType: '',
    edge: '',
    externalMemory: '',
    gprs: '',
    gps: '',
    gpu: '',
    id: '',
    imgUrl: '',
    internalMemory: [],
    model: 'Liquid Z6 Plus',
    networkSpeed: '',
    networkTechnology: '',
    nfc: '',
    options: {
      colors: [{ code: 1, name: 'Rojo' }],
      storages: [{ code: 1, name: '100GB' }],
    },
    os: '',
    price: '220',
    primaryCamera: ['13 MPautofocusLED flash'],
    radio: '',
    ram: '3 GB RAM',
    secondaryCmera: ['5 MP'],
    sensors: [],
    sim: '',
    speaker: '',
    status: '',
    usb: '',
    weight: '169',
    wlan: [],
  };
  const mockHandleClickAddProduct = jest.fn();

  const { container } = render(
    <ActionDetail product={productMock} handleClickAddProduct={mockHandleClickAddProduct} />,
  );
  expect(container).toMatchSnapshot();
});
