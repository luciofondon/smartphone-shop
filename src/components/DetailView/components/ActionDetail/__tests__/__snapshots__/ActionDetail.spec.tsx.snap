// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ActionDetail 1`] = `
<div>
  <div
    class="action-form"
  >
    <form>
      <div
        class="form-select"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="storage-label"
          >
            Almacenamiento
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="storage-label mui-component-select-storage"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-component-select-storage"
              role="button"
              tabindex="0"
            >
              100GB
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="storage"
              tabindex="-1"
              value="1"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Almacenamiento
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="form-select"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="color-label"
          >
            Color
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="color-label mui-component-select-color"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-component-select-color"
              role="button"
              tabindex="0"
            >
              Rojo
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="color"
              tabindex="-1"
              value="1"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Color
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        AÃ±adir a la cesta
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</div>
`;
